### DISCUSSION TITLE
Having trouble collecting pragma directive

### DISCUSSION THREAD

**Discussion by vsajip** [congo-cc/congo-parser-generator#157] [Q&A]
I changed a lexical definition in `CSharpLexer.ccc` from:
https://github.com/congo-cc/congo-parser-generator/blob/3d5233329e7123583df4c6b5e737729ad85cb0bb/examples/csharp/CSharpLexer.ccc#L324-L326
to
```
UNPARSED #CompilerDirective :
    <DIRECTIVE : "#" (<WS_CHAR>)* ("pragma" |  "line" | "warning" | "error" | "region" | "endregion" | "nullable") (~["\n"])+ >
;
```
but while it doesn't give any errors, something is going on that I don't understand. Given this source:
```csharp
#pragma warn disable 999
namespace foo.bar {
    /* This is a comment. */
    // This is another.
}
```
When I parse it in Java to get a `CompilationUnit` in a variable `root`, I see that `root.getSource()` is `"#namespace foo.bar  ..." ` and `root.get(0).get(0)` _is_ a `CompilerDirective`, as expected, but its `toString()` just returns `"#"` and not the whole directive. However, `root.get(0).get(0).getEndOffset()` returns 25, which is the length of the directive - but all except the first character are marked as "ignored", which is why only the first char appears. Is this some interaction with the preprocessor? I don't see why all the other chars in the directive are marked as ignored - that doesn't happen with the other unparsed tokens such as comments.

**Comment by revusky** [comment]
I guess it must have something to do with the source file locations corresponding to the preprocessor directive are marked as "ignored".

**Comment by revusky** [comment]
I'm looking at this and trying to understand what I did... A key point is [line 487, here.](https://github.com/congo-cc/congo-parser-generator/blob/main/examples/csharp/CSharpLexer.ccc#L487). Basically, the line where the \#pragma directive is is being marked as ignored so that (a bit diabolically) when you called `getSource()` to get the source corresponding to it, it is ignoring some of the characters. Actually, I am a bit confused as to why `toString()` (which delegates to `getSource() here, I guess` does not just return the empty string. But you can see that it is ignoring the characters here: https://github.com/congo-cc/congo-parser-generator/blob/3d5233329e7123583df4c6b5e737729ad85cb0bb/src/templates/java/TokenSource.java.ftl#L476

Anyway, the problem seems to be on line 487 of CSharpLexer.ccc. If you're going to just treat a \#pragma line the same as a comment, more or less, then I guess you shouldn't be marking it as ignored. So the problem would be in that token hook. But, you see, the preprocessor logic is just taking the approach that the preprocessor stuff is just ignored (along with the ignored stuff in an \#if...\#endif, say) but if you want to treat the preprocessor directives as comments that stay in the AST, then... So I guess the solution amounts to reworking that TOKEN_HOOK routine. 

Well, probably the way it deals with \#if-\#else-\#endif constructs is right, just to hide/ignore them, except for the unignored content inside them that is activated. However, if you want something like \#pragma to just be treated the same as a comment, then the `setLineSkipped` should not be called on that -- only if it is an \#if, \#elif, \#else, or \#endif. The other preprocessing directives we just treat the same as comments.

**Comment by vsajip** [comment]
Well, not everything that starts with a `#` is a preprocessor directive - those other ones are compiler directives, and should be completely ignored by the preprocessor. Anyway, that pointer to the token hook is useful, thanks - I'll see what needs changing there.

**Comment by vsajip** [comment]
The plot thickens ... it appears that this:
https://github.com/congo-cc/congo-parser-generator/blob/3d5233329e7123583df4c6b5e737729ad85cb0bb/examples/csharp/CSharpLexer.ccc#L324-L326
is masking a problem, such that source files containing `#` in random locations are parsed without errors. For example:
```csharp
#pragma warn disable 999
#
#
#

####

#
namespace foo.bar### {
#if true
    /* This is a comment. */
#else
    // This is another.
#endif
}
```
With my change to the lexer specification replacing that HASH rule, the first `#` directive is correctly parsed (apart from the line skipping, which we know needs to be changed), but a subsequent `#if` which is legal leads to an error:
```csharp
#pragma warn disable 999
namespace foo.bar {
#if true
    /* This is a comment. */
#else
    // This is another.
#endif
}
```
gives
```console
$ java CSParse reaped/dummy.cs 
Error processing file: reaped/dummy.cs
org.parsers.csharp.ParseException: 
Encountered an error at (or somewhere around) reaped/dummy.cs:3:1
Was expecting one of the following:
LBRACKET, ABSTRACT, CLASS, DELEGATE, ENUM, EXTERN, INTERFACE, INTERNAL, NEW, OVERRIDE, PRIVATE, PROTECTED, PUBLIC, READONLY, REF, SEALED, STATIC, STRUCT, UNSAFE, VIRTUAL, VOLATILE, ASYNC, PARTIAL, RECORD, REQUIRED
Found string "#" of type INVALID
	at CSharpParser.TypeDeclaration(CSharp.ccc:191)
        ...
```
I could dig into this, but you might be able to spot more quickly what the problem is.

**Comment by revusky** [comment]
I'm actually a bit puzzled about what I did in implementing the C\# preprocessor. It seems that things like \#pragma should be dealt with pretty much the same as a comment to EOL -- with the small difference that it is only valid if it occurs at the beginning of a line (modulo whitespace). I guess the way this works is something of an evolution from how the initial preprocessor I implemented worked. It shouldn't be very hard to fix this up.

**Comment by vsajip** [comment]
I'm working on the fix, but there are some weird interactions where it's not really clear why things have been done the way they have. For example, I was puzzled as to why the `#pragma` was printed as just `#`. It turns out, `setLineSkipped()` was zapping the whole line as ignored, which normally would result in an empty string, but then `cacheToken()` sets `tokenLocationTable[start_offset_of_the_token] = the_token`, which marks that particular offset as _not_ ignored, which then causes showing the `#`. The whole scheme with skipped, ignored etc, needs properly documenting (even if just by a comprehensive comment in the source code) to make maintenance easier. For example, questions such as the following should be easy to answer, but aren't (to me):
1. What happens if we _don't_ mark the #directives as ignored? What could go wrong?
2. Should we cache #directives? What happens if we don't? How would you convey in a general way (e.g. from a token hook) that a token _shouldn't_ be cached?
3. In `CSharpLexer.ccc`, error conditions in the token hook just set the token type to `INVALID` and return it, but that doesn't halt parsing and the parse completes successfully even when it shouldn't, as described in #158. However, if we return a new `InvalidToken` instead, an error _is_ raised where expected. What's the difference between a token with type `INVALID` and `InvalidToken`, and why is the latter needed if a token with type `INVALID` could do?

**Reply by revusky** [reply]
I would have thought there was no difference. Well, maybe in one case the token is marked as UNPARSED so it is ignored in any case? Well, I guess all this is not in as good shape as I thought.

**Comment by vsajip** [comment]
An interim PR is #159 - not fixing the issue yet, but making some progress towards it.

**Comment by vsajip** [comment]
Another, unrelated thing that occurs to me - I'm not sure the `PPDirectiveLine` parser is giving much bang for the buck - a whole set of parser files just for parsing a simple set of directives. Was there some particular reason the parsing of directives wasn't just built into the C# parser?

**Reply by revusky** [reply]
I was wondering the same thing myself, to be honest. It's hard to remember why I did things that way.
